daemon/network.go

```diff
@@ -28,20 +28,38 @@ func (daemon *Daemon) NetworkControllerEnabled() bool {
 	return daemon.netController != nil
 }
 
-// FindNetwork function finds a network for a given string that can represent network name or id
-func (daemon *Daemon) FindNetwork(idName string) (libnetwork.Network, error) {
-	// Find by Name
-	n, err := daemon.GetNetworkByName(idName)
-	if err != nil && !isNoSuchNetworkError(err) {
-		return nil, err
+// FindNetwork returns a network based on:
+// 1. Full ID
+// 2. Full Name
+// 3. Partial ID
+// as long as there is no ambiguity
+func (daemon *Daemon) FindNetwork(term string) (libnetwork.Network, error) {
+	listByFullName := []libnetwork.Network{}
+	listByPartialID := []libnetwork.Network{}
+	for _, nw := range daemon.GetNetworks() {
+		if nw.ID() == term {
+			return nw, nil
+		}
+		if nw.Name() == term {
+			listByFullName = append(listByFullName, nw)
+		}
+		if strings.HasPrefix(nw.ID(), term) {
+			listByPartialID = append(listByPartialID, nw)
+		}
 	}
-
-	if n != nil {
-		return n, nil
+	if len(listByFullName) == 1 {
+		return listByFullName[0], nil
 	}
-
-	// Find by id
-	return daemon.GetNetworkByID(idName)
+	if len(listByFullName) > 1 {
+		return nil, fmt.Errorf("network %s is ambiguous (%d matches found based on name)", term, len(listByFullName))
+	}
+	if len(listByPartialID) == 1 {
+		return listByPartialID[0], nil
+	}
+	if len(listByPartialID) > 1 {
+		return nil, fmt.Errorf("network %s is ambiguous (%d matches found based on ID prefix)", term, len(listByPartialID))
+	}
+	return nil, libnetwork.ErrNoSuchNetwork(term)
```

Comment:
minor nit: this might read better as a switch/case ?

```go
switch {
case len(listByFullName) == 1:
    return ...
case len(listByFullName) > 1:
    return ...
if len(listByPartialID) == 1:
    return ...
if len(listByPartialID) > 1:
    return ...

```